### PR TITLE
Rework `non_null_requirements` so it is `recursion_safe`

### LINKED ISSUE #18166: Rework `non_null_requirements` so it is `recursion_safe`
### Size estimate

L

### `LetRec` implementation (basic)

1. Descend into `(id, value)` pairs in reverse order of their definition, initializing the `gets` for each `LocalId` to `vec![]`. 
2. For non-recursive bindings, seed the `columns` for the recursive call into the binding `value` from the intersection of all sets in the context.
3. For recursive bindings seed `columns` with the empty set.

### `LetRec` implementation (advanced)

As the basic, without restriction (2).

### Notes
Maintains `Let`-based context.

### PR DESCRIPTION
**Pull request by aalexandrov**
Implements the **`LetRec` implementation (basic)** solution from #18166.

Fixes #18166.

### Motivation

  * This PR adds a known-desirable feature.

### Tips for reviewer

This is implemented similarly to #18169.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - There are no user-facing behavior changes.

### CODE REVIEW

**Comment by philip-stoev** [comment]
Item No 1. I do not think this is related to this particular PR, but relates to the more general problem of the order of the rewrites. This query:

```
EXPLAIN WITH MUTUALLY RECURSIVE c1 (f1 INTEGER, f2 INTEGER, f3 INTEGER) AS (
        SELECT 0,
               0,
               0
     UNION ALL (
                SELECT a1 . f1 + a1 . f3 + 1 AS f1,
                       a1 . f3 + 215 + 1 AS f2,
                       a1 . f2 + a1 . f1 + 1 AS f3
                  FROM c1 AS a1
                 INNER JOIN ((SELECT a1 . f2 + 254 + 1 AS f1,
                                             a1 . f3 + 80 + 1 AS f2,
                                             a1 . f3 + 1 AS f3
                                             FROM c1 AS a1 FULL OUTER JOIN c1 AS a2 USING (f2))) AS a2 USING (f1)
               )
       ) SELECT *
  FROM c1;
```

Has two different plans in this PR and the parent commit. However, if the expression `a1 . f2 + 254 + 1 ` is replaced with a1 . f2 + 255`, the plans become identical. One would naively expect that the first transformation that an optimizer does would be to fold `254+1` into `255`  . Instead, it seems that this expression persists unsimplified long enough for `non_null_requirements` to make an unexpected difference on the final pan.

**Review by philip-stoev — APPROVED** [review]
I am able to see the optimization in action. The several dozen query plans I eyeballed appear to app appear to be impacted positively.

There were no panics or wrong results.

File: src/transform/src/non_null_requirements.rs [hunk 1]
```diff
@@ -98,29 +108,59 @@ impl NonNullRequirements {
...
+                    for (id, value) in zip_eq(ids.iter().rev(), values.iter_mut().rev()) {
+                        // Compute the required non-null columns for this value.
+                        let columns = if rec_ids.contains(id) {
+                            // For recursive IDs: conservatively don't assume
+                            // any non-null column requests. TODO: This can be
+                            // improved using a fixpoint-based approximation.
+                            BTreeSet::new()
+                        } else {
+                            // For non-recursive IDs: request the intersection
+                            // of all `columns` sets in in the gets vector.
```
Comment (ggevay) [inline src/transform/src/non_null_requirements.rs:141]: in in

File: test/sqllogictest/transform/non_null_requirements.slt [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+  With Mutually Recursive
+    cte l0 =
+      Map (42) // { arity: 3 }
+        Distinct group_by=[#0, #1] // { arity: 2 }
+          Project (#0, #1) // { arity: 2 }
+            Get l0 // { arity: 3 }
+
+EOF
+
+# Multiple bindings under `WITH MUTUALLY RECURSIVE`
```
Comment (ggevay) [inline test/sqllogictest/transform/non_null_requirements.slt:50]: This doesn't seem to be affected by `non_null_requirements`. Is this intended?

**Review by ggevay — APPROVED** [review]

**Pull request merged**